core/impl/src/main/java/com/blazebit/persistence/impl/JoinManager.java

```diff
@@ -862,23 +887,35 @@ void renderValuesClausePredicate(StringBuilder sb, JoinNode rootNode, String ali
                         sb.append(attributes[j]);
                         sb.append(')');
                     } else {
+                        //alias.attributes[j]
                         sb.append(alias);
                         sb.append('.');
                         sb.append(attributes[j]);
                     }
 
+                    // alias.attributes[j] = :prefix_attributes[j]_i
                     sb.append(" = ");
 
                     sb.append(':');
                     sb.append(prefix);
                     sb.append('_');
                     sb.append(attributes[j]);
                     sb.append('_').append(i);
+
+                    if (j+1 < attributes.length) {
+                        sb.append(" AND ");
+                    }
+
+                }
+
+                if (attributes.length > 1) {
+                    sb.append(" )");
+                }
+
+                if ( i+1 < valueCount) {
                     sb.append(" OR ");
                 }
             }
-
-            sb.setLength(sb.length() - " OR ".length());
```

Comment:
I prefer a single end operation over a condition in a loop if possible. This code pattern is used all over the place.

Reply:
Hi, could you explain to me why you prefer this?

Reply:
In general, I prefer it because of performance. The if will be evaluated multiple times in the loop whereas a single end operation will be evaluated only once.

Reply:
Isn't the loop ended at the same moment and in the same way when removing `sb.setLength(sb.lenth() - " OR ".length());` and changing `if ( i+1 < valueCount)` to `if ( i < valueCount)`, as when these two things are kept the same before the above change?

I assumed your first comment meant that that you would prefer keeping the `sb.setLength(sb.lenth() - " OR ".length());` so as to avoid ending with a condition.

Reply:
Your code does
```
for (...) {
  if (condition) {
    append
  }
}
```
Whereas I prefer this
```
for (...) {
  append
}
if (!condition) {
  remove
}
```

When the loop executes 100 times, you will check the condition 100 times which I generally like to avoid if possible.

Reply:
Alright, that makes sense. Thank you.